lib/pos_fee_estimator.go

```diff
@@ -0,0 +1,159 @@
+package lib
+
+import (
+	"github.com/deso-protocol/core/collections"
+	"github.com/pkg/errors"
+	"sort"
+)
+
+type PoSFeeEstimator struct {
+	mempoolFeeEstimatorMempool    *PosMempool
+	pastBlocksTransactionRegister *TransactionRegister
+	numBlocks                     uint64
+	cachedBlocks                  []*MsgDeSoBlock
+}
+
+func (posFeeEstimator *PoSFeeEstimator) Init(mempool *PosMempool, pastBlocks []*MsgDeSoBlock, numBlocks uint64) error {
+	if mempool == nil {
+		return errors.New("PoSFeeEstimator.Init: mempool cannot be nil")
+	}
+	if !mempool.IsRunning() {
+		return errors.New("PoSFeeEstimator.Init: mempool must be running")
+	}
+	if len(pastBlocks) == 0 {
+		return errors.New("PoSFeeEstimator.Init: pastBlocks cannot be empty")
+	}
+	if numBlocks == 0 {
+		return errors.New("PoSFeeEstimator.Init: numBlocks cannot be zero")
+	}
+	if numBlocks != uint64(len(pastBlocks)) {
+		return errors.New("PoSFeeEstimator.Init: numBlocks must equal the number of pastBlocks")
+	}
+	// Sort the past blocks by height just to be safe.
+	sortedPastBlocks := collections.SortSliceStable(pastBlocks, func(ii, jj *MsgDeSoBlock) bool {
+		return ii.Header.Height < jj.Header.Height
+	})
+	minHeight := sortedPastBlocks[0].Header.Height
+	if minHeight == 0 {
+		return errors.New("PoSFeeEstimator.Init: minHeight cannot be zero")
+	}
+	// TODO: Are there any other inputs to validate?
+	posFeeEstimator.mempoolFeeEstimatorMempool = mempool
+	posFeeEstimator.numBlocks = numBlocks
+
+	// TODO: Can we get away without setting the pastBlocksUtxoView? This would
+	// save us a badger instance that isn't doing anything.
+
+	// TODO: How do we know when to update global params on the pastBlocksMempool if they're updated on the original mempool?
+	// We only really need that global params entry to help us compute the fee buckets. It's on the caller to update it
+	// using
+	// Create a mempool we can use to estimate fees for past blocks.
+	posFeeEstimator.pastBlocksTransactionRegister = NewTransactionRegister(mempool.globalParams.Copy())
+
+	// Add all the txns from the past blocks to the new pastBlocksMempool.
+	for _, block := range sortedPastBlocks {
+		if err := posFeeEstimator.addBlock(block); err != nil {
+			return errors.Wrap(err, "PosFeeEstimator.Init: error adding block to pastBlocksMempool")
+		}
+	}
+	return nil
+}
+
+func (posFeeEstimator *PoSFeeEstimator) AddBlock(block *MsgDeSoBlock) error {
+	return posFeeEstimator.addBlock(block)
```

Comment:
Wrap these public functions with a read-write mutex
```suggestion
	posFeeEstimator.lock.Lock()
	defer posFeeEstimator.lock.Unlock()

	return posFeeEstimator.addBlock(block)
```